contracts/ThreeHeapOrdering.sol

```diff
@@ -0,0 +1,307 @@
+    /// @param _heap The heap to modify.
+    /// @param _accountToShift The account to move.
+    /// @param _index The index of the account to move.
+    function shiftUp(
+        HeapArray storage _heap,
+        Account memory _accountToShift,
+        uint256 _index
+    ) private {
+        Account memory parentAccount;
+        uint256 parentIndex;
+
+        while (
+            _index > ROOT &&
+            _accountToShift.value >
+            (parentAccount = _heap.accounts[parentIndex = (_index - 1) / 3]).value
+        ) {
+            setAccount(_heap, _index, parentAccount);
+            _index = parentIndex;
+        }
+
+        setAccount(_heap, _index, _accountToShift);
+    }
+
+    /// @notice Moves an account down the heap until its value is greater than the ones of its children.
+    /// @dev This functions restores the invariant about the order of the values stored when the account to move is the only one with value smaller than what it should be.
+    /// @param _heap The heap to modify.
+    /// @param _accountToShift The account to move.
+    /// @param _index The index of the account to move.
+    function shiftDown(
+        HeapArray storage _heap,
+        Account memory _accountToShift,
+        uint256 _index
+    ) private {
+        uint256 size = _heap.size;
+        Account memory targetAccount = _accountToShift;
+        uint256 targetIndex = _index;
+        uint256 nextIndex = _index * 3;
+
+        for (;;) {
+            uint256 rightChildIndex = nextIndex + 3;
+            while (++nextIndex <= rightChildIndex && nextIndex < size) {
+                Account memory nextAccount = _heap.accounts[nextIndex];
+                if (nextAccount.value > targetAccount.value) {
+                    targetAccount = nextAccount;
+                    targetIndex = nextIndex;
+                }
+            }
+
+            if (targetIndex == _index) break;
+
+            setAccount(_heap, _index, targetAccount);
+
+            targetAccount = _accountToShift;
+            _index = targetIndex;
+            nextIndex = _index * 3;
+        }
+
+        setAccount(_heap, _index, _accountToShift);
+    }
+
+    /// @notice Inserts an account in the `_heap`.
+    /// @dev Only call this function when `_id` is not in the `_heap`.
+    /// @dev Reverts with AddressIsZero if `_value` is 0.
+    /// @param _heap The heap to modify.
+    /// @param _id The address of the account to insert.
+    /// @param _value The value of the account to insert.
+    /// @param _maxSortedUsers The maximum size of the heap.
+    function insert(
+        HeapArray storage _heap,
+        address _id,
+        uint96 _value,
+        uint256 _maxSortedUsers
+    ) private {
+        // `_heap` cannot contain the 0 address.
+        if (_id == address(0)) revert AddressIsZero();
```

Comment:
Agreed, although this feels more like an internal error (because we happen to use the address 0 to signify the absence of node). I clarified what are the sane input parameters in `update`